token/provider/external_provider_token.go

```diff
@@ -0,0 +1,209 @@
+package provider
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/fabric8-services/fabric8-auth/account"
+	"github.com/fabric8-services/fabric8-auth/application/repository"
+	"github.com/fabric8-services/fabric8-auth/errors"
+	"github.com/fabric8-services/fabric8-auth/gormsupport"
+	"github.com/fabric8-services/fabric8-auth/log"
+
+	"github.com/goadesign/goa"
+	"github.com/jinzhu/gorm"
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+)
+
+// ExternalProviderToken describes a single ExternalProviderToken
+type ExternalProviderToken struct {
+	gormsupport.Lifecycle
+	ID                   uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	ExternalProviderType uuid.UUID
+	Token                string
+	Scope                string
+	IdentityID           uuid.UUID `sql:"type:uuid"` // use NullUUID ?
+	Identity             account.Identity
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (m ExternalProviderToken) TableName() string {
+	return "external_provider_tokens"
+}
+
+// GetETagData returns the field values to use to generate the ETag
+func (m ExternalProviderToken) GetETagData() []interface{} {
+	// using the 'ID' and 'UpdatedAt' (converted to number of seconds since epoch) fields
+	return []interface{}{m.ID, strconv.FormatInt(m.UpdatedAt.Unix(), 10)}
+}
+
+// GetLastModified returns the last modification time
+func (m ExternalProviderToken) GetLastModified() time.Time {
+	return m.UpdatedAt
+}
+
+// GormExternalProviderTokenRepository is the implementation of the storage interface for
+// ExternalProviderToken.
+type GormExternalProviderTokenRepository struct {
+	db *gorm.DB
+}
+
+// NewExternalProviderTokenRepository creates a new storage type.
+func NewExternalProviderTokenRepository(db *gorm.DB) *GormExternalProviderTokenRepository {
+	return &GormExternalProviderTokenRepository{db: db}
+}
+
+// ExternalProviderTokenRepository represents the storage interface.
+type ExternalProviderTokenRepository interface {
+	repository.Exister
+	Load(ctx context.Context, id uuid.UUID) (*ExternalProviderToken, error)
+	Create(ctx context.Context, ExternalProviderToken *ExternalProviderToken) error
+	Save(ctx context.Context, ExternalProviderToken *ExternalProviderToken) error
+	Delete(ctx context.Context, id uuid.UUID) error
+	Query(funcs ...func(*gorm.DB) *gorm.DB) ([]ExternalProviderToken, error)
```

Comment:
Let's maybe add LoadByProviderAndIdentityID(ctx, providerID, identityID) method? It will simplify usage. We won't need to construct query. This method could use Query() behind the scenes or pure .Where(...).Find().

Reply:
Good idea.
1356263
Using the `Query` behind the scenes.